stages: Add stages to support ostree signatures and composefs

alexlarsson opened the pull request:
This adds two stages to support ostree ed25519 signatures (`org.osbuild.ostree.genkey` and `org.osbuild.ostree.sign`) and support for composefs (`composefs` key in `org.osbuild.ostree.config` and a new `org.osbuild.ostree.deploy-verity` stage).

Using these with a recent ostree release (2023.4) you can generate a signed ostree digest and build an image with composefs integration working.  Additionally, with the work happening in [ostree-prepare-root](https://github.com/ostreedev/ostree/pull/2921) we will have the ability to verify the generated signatures from the initrd and thus get a fully trusted boot all the way into `/usr`.

For an example of how using this may look, see: https://gitlab.com/CentOS/automotive/sample-images/-/merge_requests/333

cgwalters commented on stages/org.osbuild.ostree.deploy-verity:
```diff
@@ -0,0 +1,99 @@
+
+
+def ostree_deploy_enable_verity(deploydir):
+    cfsfile = os.path.join(deploydir, ".ostree.cfs")
```
IMO this is hardcoding an implementation detail; enabling fsverity on the objects should be owned by ostree.

cgwalters replied:
Ughhh....

Well, one argument here is that `org.osbuild.copy` should propagate the fsverity state to start.  Oh I see, it's just a thin wrapper around `cp -a` which would have to change.

So...I think what we may want here instead is something like `ostree admin deploy --post-copy` or something that does any necessary fixups after an ostree system root is copied like this.

alexlarsson replied:
> Well, one argument here is that `org.osbuild.copy` should propagate the fsverity state to start. Oh I see, it's just a thin wrapper around `cp -a` which would have to change.

Its easy to guarantee that the final target supports fs-verity, because the osbuild manifests picks fs type and runs the mkfs. However, we can't guarantee that the intermediate storage that osbuild uses supports fs-verity, so in fact there may not be anything to preserve.
 
> So...I think what we may want here instead is something like `ostree admin deploy --post-copy` or something that does any necessary fixups after an ostree system root is copied like this.

Yeah, that sounds like a reasonable approach. It could even read the fs-verity option in the config. I'll have a go at adding that.

cgwalters commented on stages/org.osbuild.ostree.genkey:
```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/python3
+"""
+Generate ed25519 public/private keypair in format used by ostree.
```
..."to bind together the initramfs and root for composefs, in a manner similar to kernel modules" or so?   Just "used by ostree" isn't very explanatory.

And let's link to the ostree composefs docs.

alexlarsson replied:
Well, technically this is just the generic "ostree sign" operation, so it has other uses too. However, you're right, we should mention the expected use case more prominently.

cgwalters replied:
OK backing up a second though...maybe this stage could land as `org.osbuild.gentransientkey` or something?  There's nothing ostree specific about it actually.

supakeen replied:
I'd prefer to keep it under `ostree`, the splitting out of the PEM is likely not relevant to everything?

cgwalters commented on stages/org.osbuild.ostree.deploy-verity:
```diff
@@ -0,0 +1,99 @@
+The way osbuild works the ostree deployment is built in a chroot and
+stored as a regular directory of files before finally being copied to
+the physical filesystem. This means ostree fs-verity support doesn't
+work, as the fs-verity setting of files is not copied.
```
Wait is this because osbuild doesn't use `ostree pull-local` but instead just copies things via e.g. `tar` or something?

I think that's what we want to fix, so that ostree is in charge of writing to the disk, not osbuild.

alexlarsson replied:
As I said above, that just doesn't work with the way osbuild is architected.

cgwalters commented on stages/org.osbuild.ostree.config:
```diff
@@ -74,6 +79,10 @@ def main(tree, options):
 
+    composefs = sysroot_options.get("composefs")
+    if composefs is not None:
+        ostree("config", "set", "ex-integrity.composefs", composefs, repo=repo)
```
So it absolutely makes sense to start this but one problem is that as soon as a PR like this is merged it becomes 10x harder to change the interface in ostree.

alexlarsson replied:
We could move this particular stage to some custom automotive specific stage. Would that help?

cgwalters replied:
cc @ondrejbudai

teg replied:
The one thing we try to avoid is two runs of the same manifest resulting in different artefacts. So we should not quietly change the behaviour of a stage.

@cgwalters the APIs used here are released in RHEL/Fedora, so presumably they already have some sort of stability guarantee? Or is this special in some way?

cgwalters replied:
The intention of the `ex.` prefix in this config option is exactly that it is experimental and subject to change.

teg replied:
Does that mean it should only be used for development, and not as part of shipped products?

cgwalters replied:
It's perhaps like Tech Preview in RHEL terms, or a nightly-only feature in Rust etc.  We need to "ship" it so it's easier to try and gain real world feedback, but not commit to supporting the exact layout and names of config options ~forever.

alexlarsson replied:
Can we maybe mirror this by calling the osbuild option `experimental-composefs`, as that is what it does, it enables an experimental feature. It will forever in the future enable an experimental feature, but you then get what you asked for.

alexlarsson replied:
One the feature stabilizes in the fututre, we can also add a regular `composefs` option.

teg replied:
Might be useful to have something mirroring this behaviour, yes. Mark something as experimental and allow it to be dropped / renamed in the future. The only thing we want to avoid is magically changing behaviour, but "experimental" manifests no longer building might be ok.

Should be up to @ondrejbudai and team though.

achilleas-k replied:
This is a whole policy change we should all agree on it seems.  I'm not entirely against it.  One thing I would say for the record is that we should never put these experimental features in other projects (osbuild/images and osbuild-composer).

In terms of what it should look like and how it should work: My first instinct is to isolate things like these into separate objects in the stage schema
```json
"properties": {
  "config": {
    "type": "object",
    "additionalProperties": false,
    "description": "OSTree configuration groups",
    ...
  },
  "experimental": {
    "composefs": {
      ...
    }
  }
}
```

(or maybe `org.osbuild.experimental` to avoid possible name collisions)
and maybe generating a warning when building a manifest that contains an experimental option.  We could even go as far as requiring a command line flag to enable it, but maybe this is a bit extreme.

cgwalters commented on stages/org.osbuild.ostree.genkey:
```diff
@@ -0,0 +1,86 @@
+#    with open('/etc/openssl.conf', 'w') as f:
+#       f.write('\n')
+
+    with tempfile.TemporaryDirectory(dir=tree) as tmpdir:
```
Since we really really want the private key to be thrown away, it probably makes sense to look at something stronger than `TemporaryDirectory`?  Maybe have openssl write to a pipe and we retain it just in memory, and then explicitly drop it?

alexlarsson replied:
The resulting secure key file is written to disk by the stage anyway, as it is then later used by the org.osbuild.ostree.sign stage. I don't see a way to pass files between stages without hitting disk anyway. However, these files are only temporarily stored on the build box, they don't get stored in any produced output.

alexlarsson commented:
In order to allow some early testing of this, I added copies of these stages (but with the org.osbuild-auto prefix) to the osbuild-auto rpm we have in the automotive team. However, long term we really want to have this in the regular osbuild, so I'm leaving this open for further discussions.

cgwalters commented:
> However, long term we really want to have this in the regular osbuild, so I'm leaving this open for further discussions.

Agreed with this!

achilleas-k commented:
> In order to allow some early testing of this, I added copies of these stages (but with the org.osbuild-auto prefix) to the osbuild-auto rpm we have in the automotive team. However, long term we really want to have this in the regular osbuild, so I'm leaving this open for further discussions.

I'm marking this as Draft as a signal that it's not ready to be merged and under discussion.  Hope that's okay.

cgwalters commented on stages/org.osbuild.ostree.sign:
```diff
@@ -0,0 +1,69 @@
+      "type": "string"
+    },
+    "ref": {
+      "description": "OStree ref to create for the commit",
```
I think this is "OSTree branch name or commit to sign"

cgwalters commented on stages/org.osbuild.ostree.sign:
```diff
@@ -0,0 +1,69 @@
+
+def ostree(*args, _input=None, **kwargs):
+    args = list(args) + [f'--{k}={v}' for k, v in kwargs.items()]
+    # Don't print this commandline, as it contains the key as an argument
```
Yeah that definitely needs to be fixed.

But ostree *does* have a shared library, which does have automatic python bindings...is there a reason those aren't being used?

supakeen replied:
It's been something I've been wanting for a while, as for now it isn't in yet though I wasn't yet aware there were Python bindings. For now it isn't in yet.

cgwalters commented on stages/org.osbuild.ostree.sign:
```diff
@@ -0,0 +1,69 @@
+    ref = options["ref"]
+    keyfile = os.path.join(tree, options["key"].lstrip("/"))
+
+    f = open(keyfile, "rb")
+    key = f.read()
+    keyid = base64.b64encode(key).decode("utf8")
```
Not that it matters here but I would tend to use the `with open` syntax because it helps avoid any GC/lifetime issues in larger fucntions.

cgwalters commented on stages/org.osbuild.ostree.sign:
```diff
@@ -0,0 +1,69 @@
+
+def ostree(*args, _input=None, **kwargs):
+    args = list(args) + [f'--{k}={v}' for k, v in kwargs.items()]
+    # Don't print this commandline, as it contains the key as an argument
```
I went to look into this, but actually we *do* support signing from a file instead of argv.  It's `--keys-file`, so we can just change to use that.

alexlarsson replied:
we now use this

alexlarsson commented:
I updated this PR with the changes that happened in ostree-auto. Main change is to make the keyfiles stored in base64 format which is the better standard for ostree use.

supakeen commented:
This discussion is still ongoing.

alexlarsson commented:
I rebased this to latest main, also updated to use the new ostree util code. I also fixed some docs according to comments here.

Outstanding issues are:
 * How should we handle the experimental aspect of enabling composefs
 * We want to move the deploy-verity stage into ostree admin and use that here

The second I will take a look at doing. The first is more a question for @cgwalters and the osbuild maintainers on how we best expose this.

alexlarsson commented:
One question though, should I split out the sign and genkey stages? They are just using supported ostree APIs and could go in as-is.

supakeen commented:
@alexlarsson let's split them out so we can get them in without all the side-quests.

achilleas-k commented on stages/org.osbuild.ostree.post-copy:
```diff
@@ -0,0 +1,50 @@
+
+def main(args, options):
+    mounts = args["mounts"]
+    root = mounts.get("root", {}).get("path")
+
+    ostree.cli("admin", "post-copy", sysroot=root)
```
It might make sense to have `sysroot` as a stage option since the names of the mounts don't really have any special meaning.  We might, for example, use mount names that are just the mountpoint (or an escaped version of it).

alexlarsson replied:
Would that just specify a mount name, or a path too?

achilleas-k replied:
just the path that you need for `ostree.cli("admin", "post-copy", sysroot=...)`

alexlarsson commented:
New version has a sysroot option

alexlarsson commented:
So, the remaining question here is, how do we want to expose this "experimental" composefs option. The proposed options seem to be, add a experimental.ostree.config stage, or add an experimental config section to the standard code stage.

With the new ostree helpers, its not really a lot of code duplication actually. So I'm personally leaning towards the experimental stage. We could even split those into their own sub-rpm (i.e. osbuild-experimental, like osbuild-ostree).

alexlarsson commented:
I went with a separate stage for this and I really like how it turned out.

alexlarsson commented:
In hindsight, I'm not 100% sure we want to split to a sub-rpm. Maybe that is a bit overkill.

supakeen commented on stages/org.osbuild.ostree.post-copy:
```diff
@@ -0,0 +1,59 @@
+after copying the final ostree tree onto the target filesystem.
+
+Note: Ensure the target filesystem supports fs-verity. See e.g.  the
+`verity` option in org.ostbuild.mkfs.ext4.
```
Minor typo.

supakeen reviewed (changes requested):
This looks good, let's *not* do the sub-RPM and instead ship experimental stages by default. I've created an issue here: https://github.com/osbuild/guides/issues/155 to document that experimental changes are subject to be dropped, schema changed, or blow up.

supakeen commented on stages/org.osbuild.ostree.post-copy:
```diff
@@ -0,0 +1,60 @@
+    }
+  }
+},
+"devices": {
```
Sorry, I think I overlooked this; but we aren't using devices anywhere in this stage are we?

alexlarsson replied:
Not directly, but the mounts don't work without devices that have the filesystems.
In sample-images for example, this is typically used like so:
```
type: org.osbuild.ostree.post-copy
devices:
  root:
    type: org.osbuild.loopback
    options:
      filename: disk.img
mounts:
- name: root
  type: org.osbuild.ext4
  source: root
  target: /
```

supakeen replied:
Thanks that makes sense.
